cli/src/semdep/parse_lockfile.py

```diff
@@ -124,6 +124,86 @@ def remove_trailing_octothorpe(s: str) -> str:
         )
 
 
+def parse_yarn2(
+    lockfile_text: str, manifest_text: Optional[str]
+) -> Iterator[FoundDependency]:
+
+    lockfile_text = "\n".join(
+        line + f" {i}" if line.strip().startswith("version") else line
+        for i, line in enumerate(lockfile_text.split("\n"))
+    )
+
+    if manifest_text:
+        manifest = json.loads(manifest_text)
+        manifest_deps = manifest["dependencies"] if "dependencies" in manifest else {}
+    else:
+        manifest_deps = None
+
+    def version_sources(line: str) -> Tuple[str, List[Tuple[str, str]]]:
+        constraint_strs = line[:-1].strip('"').split(", ")
+        constraints = []
+        for c in constraint_strs:
+            if c[0] == "@":
+                name, constraint = c[1:].split("@")
+                name = "@" + name
+            else:
+                name, constraint = c.split("@")
+            constraint = constraint.split(":")[1]  # npm:^1.0.0 --> ^1.0.0
+            constraints.append((name, constraint))
+        return constraints[0][0], constraints
+
+    def parse_dep(dep: str) -> FoundDependency:
+        lines = dep.split("\n")
+        lines = [l.strip(" ") for l in lines if len(l) - len(l.lstrip(" ")) < 4]
+        package, constraints = version_sources(lines[0])
+        field_lines = [
+            l.split(":")
+            for l in lines[1:]
+            if not (l.startswith("dependencies") or l.startswith("resolution"))
+        ]
+        fields = {f: v.strip(" ") for f, v in field_lines}
+        if "version" not in fields:
+            raise SemgrepError("yarn.lock dependency {package} missing version?")
+        version, line_number = fields["version"].split(" ")
+
+        if manifest_deps is None:
+            transitivity = Transitivity(Unknown())
+        else:
+            if package in manifest_deps:
+                transitivity = Transitivity(
+                    Direct()
+                    if (package, manifest_deps[package]) in constraints
+                    else Transitive()
+                )
+            else:
+                transitivity = Transitivity(Transitive())
+
+        return FoundDependency(
+            package=package,
+            version=version,
+            ecosystem=Ecosystem(Npm()),
+            allowed_hashes={"sha512": [fields["checksum"]]}
+            if "checksum" in fields
+            else {},
+            resolved_url=None,
+            transitivity=transitivity,
+            line_number=int(line_number),
+        )
+
+    deps = lockfile_text.split("\n\n")[2:]
+    for dep in deps:
+        yield parse_dep(dep)
+
+
+def parse_yarn(
+    lockfile_text: str, manifest_text: Optional[str]
+) -> Iterator[FoundDependency]:
+    if lockfile_text.startswith("# This file is"):
```

Comment:
Is there something in the structure of the yarn2 we can use? I feel nervous relying on the existence of a comment

Reply:
There's nothing more robust than that comment. They're all called `yarn.lock` and have similar structure. I don't think relying on a comment is bad. These files are only ever autogenerated and they always have the same comment if they're using yarn2. They always have a different comment if they're using yarn1.

Reply:
I could check for the whole comment instead of the first few words, that might be clearer?

Reply:
I guess this is fine :D